writeaheadlog.go

```diff
@@ -146,75 +145,32 @@ func newWal(path string, deps dependencies) (u []Update, w *WAL, err error) {
 	}
-
-	// When we create a new wal we don't need the caller to signal recoveryComplete
+	// No recovery needs to be performed.
 	newWal.recoveryComplete = true
-	return nil, &newWal, nil
+	return nil, newWal, nil
 }
 
 // readWALMetadata reads WAL metadata from the input file, returning an error
 // if the result is unexpected.
 func readWALMetadata(data []byte) (uint16, error) {
-	buffer := bytes.NewBuffer(data)
-
-	// Data should at least be long enough to contain the recoveryState,
-	// headerLength and versionLength
-	if buffer.Len() < 6 {
-		return 0, errors.New("data is too small")
-	}
-
-	// Prepare the variables that store the unmarshalled data
-	var recoveryState uint16
-	var headerLength uint16
-	var versionLength uint16
-
-	// Unmarshal recoveryState
-	if err := binary.Read(buffer, binary.LittleEndian, &recoveryState); err != nil {
-		return 0, build.ExtendErr("failed to unmarshal recovery state", err)
-	}
-
-	// Unmarshal header
-	if err := binary.Read(buffer, binary.LittleEndian, &headerLength); err != nil {
-		return 0, build.ExtendErr("failed to unmarshal headerLength", err)
-	}
-	if uint16(buffer.Len()) < 6+headerLength {
-		return 0, errors.New("buffer is too small to contain headerData")
-	}
-	headerData := make([]byte, headerLength)
-	if _, err := buffer.Read(headerData); err != nil {
-		return 0, build.ExtendErr("failed to read headerData", err)
-	}
-
-	// Unmarshal version
-	if err := binary.Read(buffer, binary.LittleEndian, &versionLength); err != nil {
-		return 0, build.ExtendErr("failed to unmarshal headerLength", err)
-	}
-	if uint16(buffer.Len()) < 6+headerLength+versionLength {
-		return 0, errors.New("buffer is too small to contain versionData")
-	}
-	versionData := make([]byte, versionLength)
-	if _, err := buffer.Read(versionData); err != nil {
-		return 0, build.ExtendErr("failed to read versionData", err)
+	// The metadata should at least long enough to contain all the fields.
+	if len(data) < metadataHeaderSize+metadataVersionSize+metadataStatusSize {
+		return 0, errors.New("unable to read wal metadata")
 	}
 
-	// Check the fields
-	if recoveryState != recoveryStateUnclean && recoveryState != recoveryStateWipe &&
-		recoveryState != recoveryStateClean {
-		return 0, errors.New("recoveryState didn't match")
+	// Check that the header and version match.
+	if string(data[:metadataHeaderSize]) != metadataHeader {
+		return 0, errors.New("file header is incorrect")
 	}
-	if string(headerData) != metadataHeader {
-		return 0, errors.New("header didn't match")
+	if string(data[metadataHeaderSize:metadataHeaderSize+metadataVersionSize]) != metadataVersion {
+		return 0, errors.New("file version is unrecognized - maybe you need to upgrade")
 	}
-	if string(versionData) != metadataVersion {
-		return 0, errors.New("version didn't match")
-	}
-
-	return recoveryState, nil
+	// Determine and return the current status of the file.
+	return uint16(data[metadataHeaderSize+metadataVersionSize]), nil
```

Comment:
I think you should check if the status is one of the valid ones before returning it.

Reply:
can do that